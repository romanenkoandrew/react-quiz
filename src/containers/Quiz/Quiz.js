import React, { Component } from "react";
import styles from "./Quiz.module.css";
import ActiveQuiz from "../../components/ActiveQuiz/ActiveQuiz";
import FinishedQuiz from "../../components/FinishedQuiz/FinishedQuiz";

class Quiz extends Component {
  state = {
    results: {},
    isFinished: false,
    activeQuestion: 0,
    answerState: null,
    quiz: [
      {
        id: 1,
        question: "Test",
        rightAnswerId: 2,
        answers: [
          { text: "вопрос 1", id: 1 },
          { text: "вопрос 2", id: 2 },
          { text: "вопрос 3", id: 3 },
          { text: "вопрос 4", id: 4 },
        ],
      },
      {
        id: 2,
        question: "Test2",
        rightAnswerId: 3,
        answers: [
          { text: "вопрос 1", id: 1 },
          { text: "вопрос 2", id: 2 },
          { text: "вопрос 3", id: 3 },
          { text: "вопрос 4", id: 4 },
        ],
      },
    ],
  };
  onAnswerClickHandler = (answerId) => {
    if (this.state.answerState) {
      const key = Object.keys(this.state.answerState)[0];
      if (this.state.answerState[key] === "success") {
        return;
      }
    }
    const question = this.state.quiz[this.state.activeQuestion];
    const results = this.state.results;
    if (question.rightAnswerId === answerId) {
      if (!results[question.id]) {
        results[question.id] = "success";
      }
      this.setState({ answerState: { [answerId]: "success" }, results });
      const timeout = window.setTimeout(() => {
        if (this.isQuizFinished()) {
          this.setState({ isFinished: true });
        } else {
          this.setState({
            activeQuestion: this.state.activeQuestion + 1,
            answerState: null,
          });
        }
        window.clearTimeout(timeout);
      }, 500);
    } else {
      results[question.id] = "error";
      this.setState({ answerState: { [answerId]: "error" }, results });
    }
  };
  isQuizFinished() {
    return this.state.activeQuestion + 1 === this.state.quiz.length;
  }
  retryHandler = () => {
    this.setState({
      results: {},
      isFinished: false,
      activeQuestion: 0,
      answerState: null,
    });
  };
  render() {
    return (
      <div className={styles.Quiz}>
        <div className={styles.QuizWrapper}>
          <h1>Ответьте на все вопросы</h1>
          {this.state.isFinished ? (
            <FinishedQuiz
              results={this.state.results}
              quiz={this.state.quiz}
              onRetry={this.retryHandler}
            />
          ) : (
            <ActiveQuiz
              answers={this.state.quiz[this.state.activeQuestion].answers}
              question={this.state.quiz[this.state.activeQuestion].question}
              onAnswerClick={this.onAnswerClickHandler}
              quizLength={this.state.quiz.length}
              answerNumber={this.state.activeQuestion + 1}
              state={this.state.answerState}
            />
          )}
        </div>
      </div>
    );
  }
}

export default Quiz;
