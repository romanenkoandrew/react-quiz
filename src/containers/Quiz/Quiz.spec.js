import React from "react";
import { shallow } from "enzyme";
import { Quiz } from "./Quiz";
import QuizConnect from "./Quiz";
import configureStore from "redux-mock-store";

const mockStore = configureStore();

describe("Quiz component", () => {
  let fetchQuizById, quizAnswerClick, retryQuiz;
  beforeEach(() => {
    fetchQuizById = jest.fn();
    quizAnswerClick = jest.fn();
    retryQuiz = jest.fn();
  });
  it("should render Quiz with props", () => {
    const component = shallow(
      <Quiz
        results={[]}
        quiz={[{ answers: "answers", question: "question" }]}
        isFinished={false}
        loading={false}
        activeQuestion={0}
        answerState={{ state: "state" }}
        match={{ params: { id: 0 } }}
        fetchQuizById={fetchQuizById}
        quizAnswerClick={quizAnswerClick}
        retryQuiz={retryQuiz}
      />
    );
    expect(component).toMatchSnapshot();
  });
  it("should render Quiz with loading true", () => {
    const component = shallow(
      <Quiz
        results={[]}
        quiz={[{ answers: "answers", question: "question" }]}
        isFinished={false}
        loading={true}
        activeQuestion={0}
        answerState={{ state: "state" }}
        match={{ params: { id: 0 } }}
        fetchQuizById={fetchQuizById}
        quizAnswerClick={quizAnswerClick}
        retryQuiz={retryQuiz}
      />
    );
    expect(component).toMatchSnapshot();
  });
  it("should render Quiz with isFinished true", () => {
    const component = shallow(
      <Quiz
        results={[]}
        quiz={[{ answers: "answers", question: "question" }]}
        isFinished={true}
        loading={false}
        activeQuestion={0}
        answerState={{ state: "state" }}
        match={{ params: { id: 0 } }}
        fetchQuizById={fetchQuizById}
        quizAnswerClick={quizAnswerClick}
        retryQuiz={retryQuiz}
      />
    );
    expect(component).toMatchSnapshot();
  });
  it("should call the mock retry quiz function", () => {
    const component = shallow(
      <Quiz
        results={[]}
        quiz={[{ answers: "answers", question: "question" }]}
        isFinished={true}
        loading={false}
        activeQuestion={0}
        answerState={{ state: "state" }}
        match={{ params: { id: 0 } }}
        fetchQuizById={fetchQuizById}
        quizAnswerClick={quizAnswerClick}
        retryQuiz={retryQuiz}
      />
    );
    component.unmount();
    expect(retryQuiz.mock.calls.length).toBe(1);
  });
  it("should props from store in Quiz", () => {
    const store = mockStore({
      quiz: {
        results: [],
        quiz: [{ answers: "answers", question: "question" }],
        isFinished: true,
        loading: false,
        activeQuestion: 0,
        answerState: { state: "state" },
        match: { params: { id: 0 } },
        fetchQuizById: fetchQuizById,
        quizAnswerClick: quizAnswerClick,
        retryQuiz: retryQuiz,
      },
    });
    const component2 = shallow(<QuizConnect store={store} />);
    expect(component2.children().props().results).toBeTruthy();
    expect(component2.children().props().quiz).toBeTruthy();
  });
  //   it("should roll the dice again when button is clicked", () => {
  //     const store = mockStore({
  //       quiz: {
  //         results: [],
  //         quiz: [{ answers: "answers", question: "question" }],
  //         isFinished: true,
  //         loading: false,
  //         activeQuestion: 0,
  //         answerState: { state: "state" },
  //         match: { params: { id: 0 } },
  //         fetchQuizById: fetchQuizById,
  //         quizAnswerClick: quizAnswerClick,
  //         retryQuiz: retryQuiz,
  //       },
  //     });
  //     const component2 = shallow(<QuizConnect store={store} />);
  //     component2.simulate("retryQuiz");
  //
  //     const actions = store.getActions();
  //     return console.log(component2.simulate("retryQuiz"));
  //     // expect(actions).toEqual([{ type: ActionTypes.RETRY_QUIZ }]);
  //   });
});
