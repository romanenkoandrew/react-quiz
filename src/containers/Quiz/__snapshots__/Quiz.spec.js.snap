// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz component should render Quiz with isFinished true 1`] = `
<div
  className="Quiz"
>
  <div
    className="QuizWrapper"
  >
    <h1>
      Ответьте на все вопросы
    </h1>
    <FinishedQuiz
      onRetry={[MockFunction]}
      quiz={
        Array [
          Object {
            "answers": "answers",
            "question": "question",
          },
        ]
      }
      results={Array []}
    />
  </div>
</div>
`;

exports[`Quiz component should render Quiz with loading true 1`] = `
<div
  className="Quiz"
>
  <div
    className="QuizWrapper"
  >
    <h1>
      Ответьте на все вопросы
    </h1>
    <Loader />
  </div>
</div>
`;

exports[`Quiz component should render Quiz with props 1`] = `
<div
  className="Quiz"
>
  <div
    className="QuizWrapper"
  >
    <h1>
      Ответьте на все вопросы
    </h1>
    <ActiveQuiz
      answerNumber={1}
      answers="answers"
      onAnswerClick={[MockFunction]}
      question="question"
      quizLength={1}
      state={
        Object {
          "state": "state",
        }
      }
    />
  </div>
</div>
`;
